nixos/gitea: add cache option

dadada opened the pull request:
This adds an option that sets up an in-memory cache.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

dadada committed:
nixos/gitea: add option cache

This adds an option that sets up the cache.

aanderse commented:
Personally I would either call this option `createRedisCacheLocally` or create a few options so the hard-coded values you set can be configured:
```
services.gitea.redis.createLocally = true;
# something to configure db=0,pool_size=100,idle_timeout=180 like...
services.gitea.redis.database = 0;
services.gitea.redis.poolSize = 100;
```

I think my preference is to keep it simple with `createRedisCacheLocally` and let the user configure it themselves without this option if they need more customization (noting such in the `description`).

Why do I care about the name? `useRedisCache` sounds like it should be flexible enough to use a remote host. I like using the word `Locally` to indicate this is not the case.

Aside from that everything looks great. And keep in mind the naming is just my opinion. Maybe you disagree, in which case you can feel free to ignore me 😄 

Aside from that everything looks fine from a quick glance.

ping @Ma27 for review - aren't you the maintainer of this module too? 🤔  If you're not you really should be 😄

dadada commented:
I agree. Having more flexibility is probably a good idea.

Another way might be to match `cfg.settings.cache.HOST` and configure redis, memcached or other backends supported by gitea. I'm not sure if this might be too much hidden magic though. Setting an option in `settings` should really only add a setting to the configuration file and not have additional side effects. This is why having an option outside `settings` that configures something that *just works* would be preferable, I guess.

I think exposing these settings in `services.gitea.cache` might be preferable, since gitea allows for multiple cache backends like Memcache and redis but afaik they are mutually exclusive. Caching to memory seems to be the default and does not take any extra options or configuration, so `settings` might be sufficient.

```nix
services.gitea.cache = {  
    enable = true;
    # Produces some thing like
    # HOST="redis://:macaron@127.0.0.1:6379/0?pool_size=100&idle_timeout=180s"
    "redis" =  {  # either "memory", "redis", or "memcache" with different options each
      user = "macaron"; # optional
      host = "127.0.0.1";
      port = 6379; # optional
      database = 0; # optional
      poolSize = 100; # optional
      idle_timeout = 180; # optional
    };
};
```
```
# This would produce 
# HOST="network=unix,addr=${redis.unixSocket},db=0,pool_size=100,idle_timeout=180"
services.gitea.cache = {
    enable = true;
    "redis".unixSocket = "/run/redis/redis.sock"; # mutually exclusive with (user, host, port)
}
```

For now, I'd only like to allow "redis" for the adapter, since Memcache might require additional configuration.

dadada commented:
> ```
>     "redis".unixSocket = "/run/redis/redis.sock"; # mutually exclusive with (user, host, port)
> ```

Or just `"redis".useLocal = true` and do the magic there.

dadada commented:
`services.gitea.cache.<name>.enable` should always default to false, then we can leave out `services.gitea.cache.<name>.unixSocket` and just configure the unix socket as a sane default when `services.gitea.cache."redis.enable` is set.

aanderse commented:
That sounds reasonable :+1:

Ma27 commented:
I can leave a review later tonight or tomorrow @aanderse (this also applies to the other PRs where I got CCed :)).

> ping @Ma27 for review - aren't you the maintainer of this module too? thinking If you're not you really should be smile

Since most of the modules don't have a meta.maintainers field (I don't really know about this guideline for modules tbh) I usually consider the package maintainers as responsible for the module as well unless stated otherwise... But yeah, guess I'll add myself there as well later :)

Ma27 commented on nixos/modules/services/misc/gitea.nix:
```diff
@@ -323,6 +330,15 @@ in
     services.gitea.settings = {
+
+      cache = mkMerge [
+        (mkIf cfg.useRedisCache {
```
`mkMerge` is not needed here.

Ma27 commented:
@GrahamcOfBorg test gitea

dadada commented:
Still needs some more automated tests. The redis cache already works fine using the unix socket.

hmenke commented on nixos/tests/gitea.nix:
```diff
@@ -19,6 +20,14 @@ let
           enable = true;
           database = { inherit type; };
           disableRegistration = true;
+          cache = if (cache == "redis") then {
+            redis.enable = true;
+          } else if (cache == "memcache") then {
+            memcached.enable = true;
+          } else if (cache == "memory") then {
+            memory.enable = true;
+          } else {
+          };
```
```suggestion
          cache = { ${cache}.enable = true; };
```

This actually works.  If `cache` is `null` this will evaluate to an empty attribute set:
```
nix-repl> cache = null

nix-repl> { ${cache}.enable = true; }
{ }
```

dadada committed:
nixos/gitea: clean up if-else

Co-authored-by: Henri Menke <henri@henrimenke.de>

dadada committed:
nixos/gitea: enable local redis if configured

dadada committed:
nixos/gitea: fix cache.memcached.enable

dadada committed:
nixos/gitea: add test for redis over TCP

dadada committed:
nixos/gitea: change defaults of cache.redis.port and cache.redis.host

dadada committed:
nixos/gitea: fixup redis-tcp test

dadada committed:
nixos/gitea: fix typo

dadada commented:
> Still needs some more automated tests. The redis cache already works fine using the unix socket.

Should have sufficient test coverage now. Ready for review.

dadada commented:
@GrahamcOfBorg test gitea

hmenke reviewed (approved):
I haven't tested it (I'm only using GItea with in-memory caching), but the module code looks totally sensible and there is a NixOS test. :+1:

aanderse commented on nixos/modules/services/misc/gitea.nix:
```diff
@@ -42,6 +47,81 @@ in
+          };
+        };
+        memory = {
+          enable = mkEnableOption "Enable the built-in memory cache";
```
FYI the "Enable" text is redundant and should be removed.

dadada committed:
nixos/gitea: remove redundant description

dadada committed:
nixos/gitea: update option cache.memory

Uses gitea's default cache settings or the settings provided in the
configuration.

dadada commented:
Is it possible to declare a type in a way that it can be a submodule with contents either like `cache.redis`, `cache.memcached` or `cache.memory`? Something similar to

```nix
cache = mkOption {
  type = myUnionType [ redisCache memcachedCache memory ];
};
```

where the list elements are the possible submodules.

SuperSandro2000 commented:
I think you want to construct a generic option which has a option for the cache type instead.
